pyboolnet/control_strategies.py

```diff
@@ -0,0 +1,486 @@
+
+
+def is_control_strategy(primes: dict, candidate: dict, target: List[dict], update: str, max_output: int = 1000000):
+    """
+    Check whether the *candidate* subspace is a control strategy for the *target* subset.
+
+    **arguments**:
+        * *primes*: prime implicants.
+        * *candidate*: candidate subspace.
+        * *target*: list of subspaces defining the target subset.
+        * *update*: type of update, either *"synchronous"*, *"asynchronous"* or *"mixed"*.
+        * *max_output*: maximum number of trap spaces computed. Default value: 1000000.
+
+    **returns**:
+        * *answer*: True if the *candidate* is a control strategy for the *target*, False otherwise. When False, a counterexample (minimal trap space disjoint with target or state not satisfying the CTL query) is also returned.
+
+    **example**::
+        >>> is_control_strategy(primes, {'v1': 1}, [{'v2': 0}, {'v3':1}], "asynchronous")
+    """
+
+    # Preliminary setting
+
+    if type(target) == dict:
+        print("The target must be a list of subspaces.")
+        return
+
+    if control_direct_percolation(primes, candidate, target, silent=True):
+        return True
+
+    new_primes = fix_components_and_reduce(primes, candidate, list(set([x for subs in subspaces for x in subs])))
+    tsmin = compute_trap_spaces(new_primes, "min", max_output=max_output)
+
+    # Checking control is valid in the trap spaces
+    if not control_is_valid_in_trap_spaces(new_primes, tsmin, target, update):
+        return False
+
+    # Checking general CTL query
+    return run_control_query(new_primes, target, update)
+
+
+def compute_control_strategies_with_completeness(primes: dict, target: dict, update: str = "asynchronous", limit: int = 3, avoid_nodes: List[str] = None, silent: bool = False, starting_length: int = 0, previous_cs: List[dict] = None, known_cs: List[dict] = None) -> List[dict]:
+    """
+    Identify control strategies for the *target* subspace using completeness.
+
+    **arguments**:
+        *primes*: prime implicants.
+        *target*: target subspace.
+        *update*: the type of update, either *"synchronous"*, *"asynchronous"* or *"mixed"*.
+        *limit*: maximal size of the control strategies. Default value: 3.
+        *silent*: if True, it does not print infos to screen. Default value: False.
+        *starting_length*: minimum possible size of the control strategies. Default value: 0.
+        *previous_cs*: list of already identified control strategies. Default value: empty list.
+        *avoid_nodes*: list of nodes that cannot be part of any control strategy. Default value: empty list.
+
+    **returns**:
+        * *cs_total*: list of control strategies for the *target* subspace obtained using completeness.
+
+    **example**::
+        >>> control_strategies_completeness(primes, {'v1': 1}, "asynchronous")
+
+    """
+
+    # Initializing
```

Comment:
I just noticed now, but please remove all comments that merely describe what the function names should sufficiently describe.

If you really really really want to keep a comment let me know and we'll talk about it.

Reply:
About this, I would like to use the description to link the main functions for control strategy computation (`compute_control_strategies_with_completeness` and `compute_control_strategies_with_model_checking`) to the definition of control strategy. What would be the best way to do it? I think you did something similar with some functions in the module `attractors.py`

Reply:
Hey Laura! :smiley: 

And "the definition of control strategy" is a web page? Do you mean you want to add links to the doc string? Should be explained here:
- https://sublime-and-sphinx-guide.readthedocs.io/en/latest/references.html

Reply:
Hi Hannes!
I saw that you have the comment "The approach is described and discussed in :ref:`Klarner2015(a) <klarner2015trap>`." in the functions for univocality, faithfulness and completeness and I wanted to do something similar for the functions that compute the control strategies.

Reply:
Sure, go ahead. What's your question? Just copy the string, adjust it, search for where "klarner2015trap" is defined and add your entry accordingly.

Reply:
> I just noticed now, but please remove all comments that merely describe what the function names should sufficiently describe.
> 
> If you really really really want to keep a comment let me know and we'll talk about it.

Is it okay to keep one-line comment description in each function? Should I also remove the comments describing the arguments? I tried to follow the style that you used in all the other pyboolnet functions.

Reply:
There is a difference between **doc-strings** and **comments**. Comments start with `#` and are inside the code to (supposedly) help other programmers understand your code (our remind yourself of what you intended).

Comments are bad, because you are almost certainly going to change your code and not update your comments (hence the comments will only confuse you in the future), plus you have to read the comments and compare it with the code plus it stops you from using **function names** and **variable names** to be as descriptive as possible. 

Your function names and and your variable names should describe what you're doing, they should *be the comments*.

Show me a comment you want to keep and we will replace it by better function names and better variable names. Stuff like `# initializing` and `# computing control strategies` is should just be removed.

Reply:
Ok, thanks for the clarification. I misunderstood you, it is clear now.